manim/__main__.py

```diff
@@ -46,17 +46,7 @@ def main():
         config.digest_args(args)
         input_file = config.get_dir("input_file")
 
-        if config["use_opengl_renderer"]:
-            from manim.renderer.opengl_renderer import OpenGLRenderer
-
-            for SceneClass in scene_classes_from_file(input_file):
-                try:
-                    renderer = OpenGLRenderer()
-                    scene = SceneClass(renderer)
-                    scene.render()
-                except Exception:
-                    console.print_exception()
-        elif config["use_webgl_renderer"]:
+        if config["use_webgl_renderer"]:
```

Comment:
The renderers should be initialized outside the scene in order to fit the [dependency injection](https://en.wikipedia.org/wiki/Dependency_injection#:~:text=In%20software%20engineering%2C%20dependency%20injection,object%20is%20called%20a%20service.) model.

Reply:
Thanks for the ref. If I understand dependency injection now, I assume that should be changed for both Cairo and OpenGL scenes, because as it currently stands the CairoRenderer is created in the init call too.

So I removed the default, reverted the changes in main and explicitly passed in cairo renderer when called from main, but over half the tests failed because we instantiate scenes in quite a few places (not just here).

Perhaps a fix is to define a function called get_default_renderer() that returns Cairo or OpenGL based on the flag at the time, and the default is for Scene to call that function if nothing is specified? That seems to abstract away the inner workings of the renderer from the scene.

Let me know what you think.

Reply:
Now I'm running into trouble with `camera_class` - it's set in quite a few scene types and it looks like it's a value passed right into the renderer. So the scene needs to know what kind of renderer it has in order to pass the camera to the renderer (with Cairo) or not (with OpenGL).

Because this is a [relatively] easy fix, I'm going to assume the camera class should just be passed transparently on to the renderer - the scene doesn't even need to remember it. Therefore, I've added camera_class as an argument to OpenGLRenderer, but it gives an error if anything other than None or OpenGLCamera is passed to it.

Reply:
You don't have to worry about doing the same for the CairoRenderer. It would be nice, but there are way too many places where manim behaves inconsistently to address them all. I just want to avoid this particular regression with the OpenGLRenderer.

Reply:
Does the code as written (get_default_renderer, adding the camera_class argument to OpenGLRenderer) work?

Reply:
If not, I can pull it back, but then I'd have to have a flag in GraphicalUnitTest to tell the Scenes in the OpenGL test to use OpenGL, which doesn't feel like the right place to do that, or that flexible of an option.

Reply:
As written it doesn't fit the dependency injection model; the renderer can only be instantiated with data from the scene.

Using a flag (or updating the config and reading config["use_opengl_renderer"]) seems fine to me. I've been using that line throughout the codebase anyway.

Reply:
Resolved - see changes in 0dac717 and 2f6c276, but the best view is [the overall change log](https://github.com/ManimCommunity/manim/pull/1160/files).